### PR TITLE
Issue 1369: Correct segment metrics

### LINKED ISSUE #1369: Segment Metrics being recorded as Gauge and Counter
**Problem description**
While working with the auto scaling, we observed that the segments_count, segment_merges and segment_splits are all being recorded with a Gauge and a Counter.

screen shot 2017-05-31 at 8 35 58 am

The actual data is being stored under:
segments_count.Counter
segment_merges.Gauge
segment_splits.Gauge

The expectation is that the data is stored under:
segments_count.Gauge
segment_merges.Counter
segment_splits.Counter

**Problem location**
Metrics

**Suggestions for an improvement**
Store metrics under:
segments_count.Gauge
segment_merges.Counter
segment_splits.Counter

Where Gauge has the current number of segment and the Counters are running totals of the number of splits/merges.

### PR DESCRIPTION
**Pull request by Abhijeet-Jadhav**
**Change log description**
Corrected the reporting of segment metrics.

**Purpose of the change**
Fixes #1369 

**What the code does**
Corrects the value types associated.

**How to verify it**
startStandlone gradle task creates a few metrics in /tmp/csv/pravega.

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java [hunk 1]
```diff
@@ -334,14 +334,19 @@ public OperationContext createContext(String scope, String name) {
                 new AbstractMap.SimpleEntry<>(x.getKeyStart(), x.getKeyEnd())).collect(Collectors.toList());
 
         future.thenAccept(result -> {
-            DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_COUNT, scope, name),
+            DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_COUNT, scope, name),
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: this could be a negative number. is that ok for SEGMENTS_COUNT?
also, i dont get this logic -- this metric is a gauge as well as a counter. 
I dont see any reason for it to be counter though.. gauge makes more sense.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java [hunk 2]
```diff
@@ -334,14 +334,19 @@ public OperationContext createContext(String scope, String name) {
...
                     newSegments.size() - sealedSegments.size());
             getSealedRanges(scope, name, sealedSegments, context, executor)
                     .thenAccept(sealedRanges -> {
-                        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_SPLITS, scope, name),
-                                findSplits(sealedRanges, newRanges));
-                        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_MERGES, scope, name),
-                                findSplits(newRanges, sealedRanges));
+                        long oldNumSplits = DYNAMIC_LOGGER.getCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, name));
+                        long newNumSplits = findSplits(sealedRanges, newRanges);
+                        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, name),
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: SEGMENTS_SPLITS should also be a gauge. i dont see how this is a counter as we are not accumulating values here.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java [hunk 3]
```diff
@@ -334,14 +334,19 @@ public OperationContext createContext(String scope, String name) {
...
+                                    (newNumSplits > oldNumSplits) ? (newNumSplits - oldNumSplits) : newNumSplits );
+
+                        long oldNumMerges = DYNAMIC_LOGGER.getCounterValue(nameFromStream(SEGMENTS_MERGES, scope, name));
+                        long newNumMerges = findSplits(newRanges, sealedRanges);
+                        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_MERGES, scope, name),
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: same as above

File: shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java [hunk 4]
```diff
@@ -95,6 +95,24 @@ public Counter call() throws Exception {
...
+        Exceptions.checkNotNullOrEmpty(name, "name");
+        String counterName = name + ".Counter";
+        try {
+            Counter counter = countersCache.get(counterName, new Callable<Counter>() {
+                @Override
+                public Counter call() throws Exception {
+                    return underlying.createCounter(counterName);
+                }
+            });
+            return counter.get();
```
Comment (shiveshr) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: if there is a cache miss, we are effectively resetting the counter.
this can lead to incorrect computations.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java [hunk 5]
```diff
@@ -334,14 +334,19 @@ public OperationContext createContext(String scope, String name) {
...
-                                findSplits(sealedRanges, newRanges));
-                        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_MERGES, scope, name),
-                                findSplits(newRanges, sealedRanges));
+                        long oldNumSplits = DYNAMIC_LOGGER.getCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, name));
+                        long newNumSplits = findSplits(sealedRanges, newRanges);
+                        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, name),
+                                    (newNumSplits > oldNumSplits) ? (newNumSplits - oldNumSplits) : newNumSplits );
+
+                        long oldNumMerges = DYNAMIC_LOGGER.getCounterValue(nameFromStream(SEGMENTS_MERGES, scope, name));
+                        long newNumMerges = findSplits(newRanges, sealedRanges);
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: shouldnt this method be findMerges

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 2 above)
Comment (bayarsaikhan) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: @shiveshr It think number of splits can be counter. Gauge values are somewhat inconsistent and not based on historic value. Counter seems to be working better. Please confirm @glounder

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 5 above)
Comment (bayarsaikhan) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: This may be leading to incorrect results. Should be findMerges or similar.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 1 above)
Comment (bayarsaikhan) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: I think that is what we want. Please confirm @glounder

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 2 above)
Comment (glounder) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: For the splits and merges, we want to know the rate of when they happen.  (eg.  Show the number of splits per hour).  There is no concept of the "current number of splits" because its an event that happens.  Because of this, a gauge doesn't make sense.  We need the counter to calculate the rate of when these things occur.

For Segment Counts, that makes sense as a Gauge, because at any point in time, we'd want to know how many segments do we currently have.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 1 above)
Comment (glounder) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: I was expecting the segments_count metric to reflect the current number of segments for stream at any point in time.  If you are increasing/decreasing the value over time, that's ok (eg subtract 2 from current count), but I'm assuming that the current count wouldn't go below 1.

Comment (fpj) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: @shiveshr in which situation would the segment count be negative? It should reflect the instantaneous value corresponding to the number of segments of the stream, no?

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 2 above)
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: The code actually has the following, which sounds right to me:
```
   // Stream segment counts (Dynamic)
    public static final String SEGMENTS_COUNT = "controller.segments_count";   // Dynamic Gauge
    public static final String SEGMENTS_SPLITS = "controller.segment_splits"; // Dynamic Counter
    public static final String SEGMENTS_MERGES = "controller.segment_merges"; // Dynamic Counter
```

splits should be the accumulated number of splits, yes?

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 1 above)
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: this is doing 'newSegments.size() - sealedSegments.size()'
so if we seal segments {0,1} and create segments {2} then this value will be -1. 

from glounder's comment it seems the expectation is active segment count. which is not what the values here seem to provide.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 2 above)
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: ok so this is overall accumulated count of splits. but i dont quite follow the logic here -- (newNumSplits > oldNumSplits) ? (newNumSplits - oldNumSplits) : newNumSplits );

this does not seem like accumulation but like a gauge value.

Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: @shiveshr The method incCounterValue takes input a delta, hence adding the delta value to counter. I am just making sure we are not adding negative values to splits, since it is accumulated number of splits. If oldNumSplits > newNumSplits just adding the newNumSplits value.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 1 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: yes, it should be just `newSegments.size()`, right @fpj @shiveshr ?

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 3 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: As commented by @glounder above, for splits/merges gauge doesn't make sense.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 5 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: This logic (use of findSplits) was already existing in code. This PR addresses the correctness of metrics' types reported.
Basically, findSplits is reused for findMerges by interchanging the parameters. But I agree, the name is misleading, will change it to have generic name for such method like findDeltaNumSegments.

File: shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java (hunk 4 above)
Comment (Abhijeet-Jadhav) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: Yes, we are storing the dynamic metrics' values in cache. If there is cache miss (value doesn't exist or value is not present due to cache eviction ) then we are resetting the counter.  Correct me if I am wrong @fpj .

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 5 above)
Comment (bayarsaikhan) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: The name is misleading. A quick change of name, or addition of a new method would help future developers avoid further confusions

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 1 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: Done.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 5 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: Done.

File: shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java (hunk 4 above)
Comment (fpj) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: That's right because this metrics code doesn't have access to anything else, so if it is not in its local cache, then it will be reset. I don't think there is a way around it, unless we add some callback mechanism to enable the metrics code to request an uncached value. What do you guys think?

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java [hunk 6]
```diff
@@ -334,14 +334,18 @@ public OperationContext createContext(String scope, String name) {
                 new AbstractMap.SimpleEntry<>(x.getKeyStart(), x.getKeyEnd())).collect(Collectors.toList());
 
         future.thenAccept(result -> {
-            DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_COUNT, scope, name),
-                    newSegments.size() - sealedSegments.size());
+            DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_COUNT, scope, name), newSegments.size());
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: new segments.size is incorrect... newsegments == new segments created in this scale operation.. and not overall segments.. 
for overall segments you need to get active segments.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java [hunk 7]
```diff
@@ -334,14 +334,18 @@ public OperationContext createContext(String scope, String name) {
...
-                                findSplits(sealedRanges, newRanges));
-                        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_MERGES, scope, name),
-                                findSplits(newRanges, sealedRanges));
+                        long oldNumSplits = DYNAMIC_LOGGER.getCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, name));
+                        long newNumSplits = findDeltaNumSplits(sealedRanges, newRanges);
+                        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, name),
+                                    (newNumSplits > oldNumSplits) ? (newNumSplits - oldNumSplits) : newNumSplits );
+
+                        long oldNumMerges = DYNAMIC_LOGGER.getCounterValue(nameFromStream(SEGMENTS_MERGES, scope, name));
+                        long newNumMerges = findDeltaNumSplits(newRanges, sealedRanges);
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: findDeltaNumMerges?

File: shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java (hunk 4 above)
Comment (shiveshr) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: yes, but in such cases, since we have the record in the metadata, we can always compute the potential previous value of the counter and update it correctly

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 7 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: Oh, I wanted to name it is as findDeltaNumSegments so it can be used for both splits/merges. Will rename again.

Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: no i was not talking about the rename.. what you have is fine.

File: shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java [hunk 8]
```diff
@@ -95,6 +95,24 @@ public Counter call() throws Exception {
...
 
     @Override
+    public long getCounterValue(String name) {
+        Exceptions.checkNotNullOrEmpty(name, "name");
+        String counterName = name + ".Counter";
+        try {
+            Counter counter = countersCache.get(counterName, new Callable<Counter>() {
+                @Override
+                public Counter call() throws Exception {
+                    return underlying.createCounter(counterName);
```
Comment (fpj) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: Is this behavior correct, that we create a counter in the case it doesn't exist?

Comment (Abhijeet-Jadhav) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: Yes, it seems correct. The get call's signature `get(K key, Callable<? extends V> valueLoader)` expects callable. https://google.github.io/guava/releases/15.0/api/docs/com/google/common/cache/Cache.html#get(K, java.util.concurrent.Callable)

Comment (shiveshr) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: this is a loading cache, in case the value is not present in cache, it uses the supplied loader to load the value.. 
the problem here, though, is that loader is simply creating a new counter which will start with count 0. 
in another comment i had asked abhijit to ensure that a lambda be passed which will be used to compute the correct value and load it in cache.

Comment (fpj) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: I think my comment wasn't clear, so let me try again. `getCounterValue` is supposed to return a counter value and getters typically do not have side-effects. I understand that the `get` call in the cache enables the behavior implemented with the `Callable`, but I'm not sure we should be making use of it here. Could you convince me that the expected behavior is to create a counter with zero value or populate it with the lambda that @shiveshr suggests?

Comment (Abhijeet-Jadhav) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: As pointed out correctly, we shouldn't have a side effect for getCounterValue. Will remove this in next commit.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java [hunk 9]
```diff
@@ -334,14 +334,18 @@ public OperationContext createContext(String scope, String name) {
...
-            DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_COUNT, scope, name),
-                    newSegments.size() - sealedSegments.size());
+            DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_COUNT, scope, name), newSegments.size());
             getSealedRanges(scope, name, sealedSegments, context, executor)
                     .thenAccept(sealedRanges -> {
-                        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_SPLITS, scope, name),
-                                findSplits(sealedRanges, newRanges));
-                        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_MERGES, scope, name),
-                                findSplits(newRanges, sealedRanges));
+                        long oldNumSplits = DYNAMIC_LOGGER.getCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, name));
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: another aspect to note here is -- getCounterValue from cache may be incorrect because 'manual' scale requests can be received on different controller instances.. they may each build their own cache and these counted values may diverge and we will have incorrect counted values on different controller instances.. 

so as a solution, we can always compute this value in context of this scale operation (as we have the metadata already read from the store) and not use a cache for these two counters. 
it will take care of both the problems of cache miss getting reset and computing incorrect values.

File: shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java (hunk 8 above)
Comment (shiveshr) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: IMO get should not have a sideeffect on metadata.. this is cache we are talking about.. in case of cache miss, get should retrieve and load the value.. the important distinction to note is that this get is not updating the metadata anywhere for this get to have any side effect..

File: shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java (hunk 4 above)
Comment (shiveshr) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: in one of the comments above, i have suggested not to use cached values for certain cases where different requests may be handled by different controller instances and each having its own cache which can diverge. in such cases we should compute value and report it rather than pick it from the cache.

File: shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java (hunk 8 above)
Comment (fpj) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: yeah, exactly. we need to define the behavior we want before we decide on the implementation. @shiveshr is right that a cache will often load an existing value in the case the value isn't in the case, so the logic is:

```
get (Key k):
   if( k in the cache ) then
        return value of k
   else
        load (k, v)
        return value of k
```

This is a get on the cache, though. A higher level get call (e.g., `getCounterValue`) that invokes the cache get might not want to do the whole set up for a counter metric if it is not there. Can we define the behavior first and then agree on the implementation?

Comment (Abhijeet-Jadhav) [inline shared/metrics/src/main/java/io/pravega/shared/metrics/DynamicLoggerImpl.java]: @shivesh suggests to make use gauge for splits/merges metrics and compute the value everytime we hit this part of code.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java (hunk 9 above)
Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: Ok, we can use gauge. But we need to cache the values anyway for eviction of this dynamic metric. 
We won't use the value from cache, but we will use cache for eviction, as has been done for other gauge metrics.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java [hunk 10]
```diff
@@ -334,15 +334,15 @@ public OperationContext createContext(String scope, String name) {
...
-            DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_COUNT, scope, name),
-                    newSegments.size() - sealedSegments.size());
-            getSealedRanges(scope, name, sealedSegments, context, executor)
-                    .thenAccept(sealedRanges -> {
-                        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_SPLITS, scope, name),
-                                findSplits(sealedRanges, newRanges));
-                        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_MERGES, scope, name),
-                                findSplits(newRanges, sealedRanges));
-                    });
+            getActiveSegments(scope, name, System.currentTimeMillis(), null, null).thenAccept(list -> {
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: this returns a future. you can chain it to previous result by doing thenCompose and in that return Futures.allOf(getActiveSegments.chain, findNumSplits.chain, findNumMerges.chain)

Comment (Abhijeet-Jadhav) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java]: Done.
